charmcraft/charm_builder.py

```diff
@@ -251,44 +251,36 @@ def _install_dependencies(self, staging_venv_dir):
                 self._install_strict_dependencies(pip_cmd)
                 return
 
-            # Legacy non-strict dependencies.
-            # This method is not valid for any bases added after 2024-01-01 or for DEVEL bases.
-            try:
+            # Non-strict dependency resolution:
+            # 1. Install binary-allowed packages
+            # 2. Install source packages
+            # 3. Install from requirements files and charm libs dependencies
+            if self.binary_python_packages:
+                print(
+                    "Installing binary-allowed packages and their dependencies.\n"
+                    "WARNING: dependencies may also be installed from binary wheels."
+                )
                 _process_run(
                     get_pip_command(
                         [pip_cmd, "install"],
-                        self.requirement_paths,
-                        source_deps=[*self.python_packages, *self.charmlib_deps],
+                        [],
```

Comment:
pedantic: the previous usage hinted at what this parameter is.  Now that it is an empty list, it is not obvious.
```suggestion
                        requirement_files=[],
```

Reply:
Good point. Added!